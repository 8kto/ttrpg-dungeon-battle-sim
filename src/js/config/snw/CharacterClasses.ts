import { Dice } from '../../domain/Dice'
import { Alignment, AttrScore, CharacterClass, CharacterClassDef, CharacterRace } from '../../domain/snw/CharacterClass'

export const ANY_RACE = Object.values(CharacterRace) as CharacterRace[]

export const ANY_ALIGNMENT = Object.values(Alignment) as Alignment[]

export const PRIME_ATTR_MIN = 13

const AssassinClassDef: CharacterClassDef = {
  name: CharacterClass.Assassin,
  PrimeAttr: [
    [AttrScore.Strength, PRIME_ATTR_MIN],
    [AttrScore.Dexterity, PRIME_ATTR_MIN],
    [AttrScore.Intelligence, PRIME_ATTR_MIN],
  ],
  StrictAttr: [
    [AttrScore.Strength, 12],
    [AttrScore.Dexterity, 12],
    [AttrScore.Intelligence, 12],
  ],
  HitDice: Dice.d6,
  ArmorPermitted: 'Leather armor only, shield permitted',
  WeaponsPermitted: 'Any',
  Race: [CharacterRace.Human],
  Alignment: [Alignment.Neutral, Alignment.Chaotic],
}

const ClericClassDef: CharacterClassDef = {
  name: CharacterClass.Cleric,
  PrimeAttr: [[AttrScore.Wisdom, PRIME_ATTR_MIN]],
  HitDice: Dice.d6,
  ArmorPermitted: 'Any',
  WeaponsPermitted:
    'Blunt weapons only (club, flail, hammer, mace, staff, etc.). ' +
    'No missile weapons other than oil or slings if the Referee permits.',
  Race: [CharacterRace.Human, CharacterRace.HalfElf],
  Alignment: [Alignment.Lawful, Alignment.Chaotic],
  $isCaster: true,
  $spellsAtTheFirstLevel: 0,
}

const DruidClassDef: CharacterClassDef = {
  name: CharacterClass.Druid,
  PrimeAttr: [
    [AttrScore.Wisdom, PRIME_ATTR_MIN],
    [AttrScore.Charisma, PRIME_ATTR_MIN],
  ],
  StrictAttr: [
    [AttrScore.Wisdom, 12],
    [AttrScore.Charisma, 14],
  ],
  HitDice: Dice.d6,
  ArmorPermitted: 'Leather armor, wooden shield',
  WeaponsPermitted: 'Dagger, sickle-shaped sword (treat as short sword), spear, sling, oil',
  Race: [CharacterRace.Human],
  Alignment: [Alignment.Neutral],
  $isCaster: true,
  $spellsAtTheFirstLevel: 1,
}

const FighterClassDef: CharacterClassDef = {
  name: CharacterClass.Fighter,
  PrimeAttr: [[AttrScore.Strength, PRIME_ATTR_MIN]],
  HitDice: Dice.d8,
  ArmorPermitted: 'Any',
  WeaponsPermitted: 'Any',
  Race: ANY_RACE,
  Alignment: ANY_ALIGNMENT,
}

const MagicUserClassDef: CharacterClassDef = {
  name: CharacterClass.MagicUser,
  PrimeAttr: [[AttrScore.Intelligence, PRIME_ATTR_MIN]],
  HitDice: Dice.d4,
  ArmorPermitted: 'None',
  WeaponsPermitted: 'Dagger, staff, and darts',
  Race: [CharacterRace.Elf, CharacterRace.HalfElf, CharacterRace.Human],
  Alignment: ANY_ALIGNMENT,
  $isCaster: true,
  $spellsAtTheFirstLevel: 1,
}

const MonkClassDef: CharacterClassDef = {
  name: CharacterClass.Monk,
  PrimeAttr: [[AttrScore.Wisdom, PRIME_ATTR_MIN]],
  StrictAttr: [
    [AttrScore.Strength, 12],
    [AttrScore.Dexterity, 15],
    [AttrScore.Wisdom, 15],
  ],
  HitDice: Dice.d4,
  ArmorPermitted: 'None',
  WeaponsPermitted: 'Any',
  Race: [CharacterRace.Human],
  Alignment: ANY_ALIGNMENT,
}

const PaladinClassDef: CharacterClassDef = {
  name: CharacterClass.Paladin,
  PrimeAttr: [[AttrScore.Strength, PRIME_ATTR_MIN]],
  StrictAttr: [[AttrScore.Charisma, 17]],
  HitDice: Dice.d8,
  ArmorPermitted: 'Any',
  WeaponsPermitted: 'Any',
  Race: [CharacterRace.Human],
  Alignment: [Alignment.Lawful],
}

const RangerClassDef: CharacterClassDef = {
  name: CharacterClass.Ranger,
  PrimeAttr: [[AttrScore.Strength, PRIME_ATTR_MIN]],
  StrictAttr: [
    [AttrScore.Constitution, 15],
    [AttrScore.Intelligence, 12],
    [AttrScore.Wisdom, 12],
  ],
  HitDice: Dice.d8,
  ArmorPermitted: 'Any',
  WeaponsPermitted: 'Any',
  Race: [CharacterRace.Human],
  Alignment: [Alignment.Lawful],
  $meta: '2d8 at first level, 1d8/level thereafter',
}

const ThiefClassDef: CharacterClassDef = {
  name: CharacterClass.Thief,
  PrimeAttr: [[AttrScore.Dexterity, PRIME_ATTR_MIN]],
  HitDice: Dice.d4,
  ArmorPermitted: 'Leather armor only; no shield',
  WeaponsPermitted: 'All, but magical weapons are limited to daggers and swords',
  Race: ANY_RACE,
  Alignment: [Alignment.Neutral, Alignment.Chaotic],
}

/**
 * Record for Character Class Definitions
 */
export const CharacterClasses: Record<CharacterClass, CharacterClassDef> = {
  [CharacterClass.Assassin]: AssassinClassDef,
  [CharacterClass.Cleric]: ClericClassDef,
  [CharacterClass.Druid]: DruidClassDef,
  [CharacterClass.Fighter]: FighterClassDef,
  [CharacterClass.MagicUser]: MagicUserClassDef,
  [CharacterClass.Monk]: MonkClassDef,
  [CharacterClass.Paladin]: PaladinClassDef,
  [CharacterClass.Ranger]: RangerClassDef,
  [CharacterClass.Thief]: ThiefClassDef,
}
