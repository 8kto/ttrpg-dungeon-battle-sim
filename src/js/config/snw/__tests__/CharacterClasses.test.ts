import { CharacterClasses } from '../CharacterClasses'

describe('characterClasses', () => {
  it('should match the snapshot', () => {
    expect(CharacterClasses).toMatchSnapshot()
  })
})
