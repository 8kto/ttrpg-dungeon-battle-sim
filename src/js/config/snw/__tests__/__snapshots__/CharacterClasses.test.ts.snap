// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`characterClasses should match the snapshot 1`] = `
{
  "Assassin": {
    "Alignment": [
      "Neutral",
      "Chaotic",
    ],
    "ArmorPermitted": "Leather armor only, shield permitted",
    "HitDice": 6,
    "PrimeAttr": [
      [
        "Strength",
        13,
      ],
      [
        "Dexterity",
        13,
      ],
      [
        "Intelligence",
        13,
      ],
    ],
    "Race": [
      "Human",
    ],
    "SavingThrow": {
      "alternative": {
        "DeathRaysAndPoison": 13,
        "DragonsBreath": 16,
        "SpellsAndStaffs": 15,
        "TurnedToStone": 13,
        "Wands": 14,
      },
      "snw": {
        "details": null,
        "value": 15,
      },
    },
    "StrictAttr": [
      [
        "Strength",
        12,
      ],
      [
        "Dexterity",
        12,
      ],
      [
        "Intelligence",
        12,
      ],
    ],
    "WeaponsPermitted": "Any",
    "name": "Assassin",
  },
  "Cleric": {
    "$isCaster": true,
    "$spellsAtTheFirstLevel": 0,
    "Alignment": [
      "Lawful",
      "Chaotic",
    ],
    "ArmorPermitted": "Any",
    "HitDice": 6,
    "PrimeAttr": [
      [
        "Wisdom",
        13,
      ],
    ],
    "Race": [
      "Human",
      "HalfElf",
    ],
    "SavingThrow": {
      "alternative": {
        "DeathRaysAndPoison": 11,
        "DragonsBreath": 16,
        "SpellsAndStaffs": 15,
        "TurnedToStone": 14,
        "Wands": 12,
      },
      "snw": {
        "details": "+2 against being paralyzed or poisoned",
        "value": 15,
      },
    },
    "WeaponsPermitted": "Blunt weapons only (club, flail, hammer, mace, staff, etc.). No missile weapons other than oil or slings if the Referee permits.",
    "name": "Cleric",
  },
  "Druid": {
    "$isCaster": true,
    "$spellsAtTheFirstLevel": 1,
    "Alignment": [
      "Neutral",
    ],
    "ArmorPermitted": "Leather armor, wooden shield",
    "HitDice": 6,
    "PrimeAttr": [
      [
        "Wisdom",
        13,
      ],
      [
        "Charisma",
        13,
      ],
    ],
    "Race": [
      "Human",
    ],
    "SavingThrow": {
      "alternative": {
        "DeathRaysAndPoison": 11,
        "DragonsBreath": 16,
        "SpellsAndStaffs": 15,
        "TurnedToStone": 14,
        "Wands": 12,
      },
      "snw": {
        "details": "+2 bonus against fire",
        "value": 15,
      },
    },
    "StrictAttr": [
      [
        "Wisdom",
        12,
      ],
      [
        "Charisma",
        14,
      ],
    ],
    "WeaponsPermitted": "Dagger, sickle-shaped sword (treat as short sword), spear, sling, oil",
    "name": "Druid",
  },
  "Fighter": {
    "Alignment": [
      "Chaotic",
      "Neutral",
      "Lawful",
    ],
    "ArmorPermitted": "Any",
    "HitDice": 8,
    "PrimeAttr": [
      [
        "Strength",
        13,
      ],
    ],
    "Race": [
      "Human",
      "Dwarf",
      "Elf",
      "HalfElf",
      "Hobbit",
    ],
    "SavingThrow": {
      "alternative": {
        "DeathRaysAndPoison": 12,
        "DragonsBreath": 15,
        "SpellsAndStaffs": 16,
        "TurnedToStone": 14,
        "Wands": 13,
      },
      "snw": {
        "details": "+1 bonus on all saving throws except against spells (including those from wands and staffs)",
        "value": 15,
      },
    },
    "WeaponsPermitted": "Any",
    "name": "Fighter",
  },
  "MagicUser": {
    "$isCaster": true,
    "$spellsAtTheFirstLevel": 1,
    "Alignment": [
      "Chaotic",
      "Neutral",
      "Lawful",
    ],
    "ArmorPermitted": "None",
    "HitDice": 4,
    "PrimeAttr": [
      [
        "Intelligence",
        13,
      ],
    ],
    "Race": [
      "Elf",
      "HalfElf",
      "Human",
    ],
    "SavingThrow": {
      "alternative": {
        "DeathRaysAndPoison": 13,
        "DragonsBreath": 16,
        "SpellsAndStaffs": 15,
        "TurnedToStone": 13,
        "Wands": 14,
      },
      "snw": {
        "details": "+2 against spells, including spells from magic wands and staffs",
        "value": 15,
      },
    },
    "WeaponsPermitted": "Dagger, staff, and darts",
    "name": "MagicUser",
  },
  "Monk": {
    "Alignment": [
      "Chaotic",
      "Neutral",
      "Lawful",
    ],
    "ArmorPermitted": "None",
    "HitDice": 4,
    "PrimeAttr": [
      [
        "Wisdom",
        13,
      ],
    ],
    "Race": [
      "Human",
    ],
    "SavingThrow": {
      "alternative": {
        "DeathRaysAndPoison": 11,
        "DragonsBreath": 16,
        "SpellsAndStaffs": 15,
        "TurnedToStone": 14,
        "Wands": 12,
      },
      "snw": {
        "details": "+2 bonus against paralysis and poisons (when Wisdom is 12+)",
        "value": 15,
      },
    },
    "StrictAttr": [
      [
        "Strength",
        12,
      ],
      [
        "Dexterity",
        15,
      ],
      [
        "Wisdom",
        15,
      ],
    ],
    "WeaponsPermitted": "Any",
    "name": "Monk",
  },
  "Paladin": {
    "Alignment": [
      "Lawful",
    ],
    "ArmorPermitted": "Any",
    "HitDice": 8,
    "PrimeAttr": [
      [
        "Strength",
        13,
      ],
    ],
    "Race": [
      "Human",
    ],
    "SavingThrow": {
      "alternative": {
        "DeathRaysAndPoison": 12,
        "DragonsBreath": 15,
        "SpellsAndStaffs": 16,
        "TurnedToStone": 14,
        "Wands": 13,
      },
      "snw": {
        "details": null,
        "value": 12,
      },
    },
    "StrictAttr": [
      [
        "Charisma",
        17,
      ],
    ],
    "WeaponsPermitted": "Any",
    "name": "Paladin",
  },
  "Ranger": {
    "$meta": "2d8 at first level, 1d8/level thereafter",
    "Alignment": [
      "Lawful",
    ],
    "ArmorPermitted": "Any",
    "HitDice": 8,
    "PrimeAttr": [
      [
        "Strength",
        13,
      ],
    ],
    "Race": [
      "Human",
    ],
    "SavingThrow": {
      "alternative": {
        "DeathRaysAndPoison": 12,
        "DragonsBreath": 15,
        "SpellsAndStaffs": 16,
        "TurnedToStone": 14,
        "Wands": 13,
      },
      "snw": {
        "details": null,
        "value": 14,
      },
    },
    "StrictAttr": [
      [
        "Constitution",
        15,
      ],
      [
        "Intelligence",
        12,
      ],
      [
        "Wisdom",
        12,
      ],
    ],
    "WeaponsPermitted": "Any",
    "name": "Ranger",
  },
  "Thief": {
    "Alignment": [
      "Neutral",
      "Chaotic",
    ],
    "ArmorPermitted": "Leather armor only; no shield",
    "HitDice": 4,
    "PrimeAttr": [
      [
        "Dexterity",
        13,
      ],
    ],
    "Race": [
      "Human",
      "Dwarf",
      "Elf",
      "HalfElf",
      "Hobbit",
    ],
    "SavingThrow": {
      "alternative": {
        "DeathRaysAndPoison": 13,
        "DragonsBreath": 16,
        "SpellsAndStaffs": 15,
        "TurnedToStone": 13,
        "Wands": 14,
      },
      "snw": {
        "details": "+2 against devices, including traps, magical wands or staffs, and other magical devices",
        "value": 15,
      },
    },
    "WeaponsPermitted": "All, but magical weapons are limited to daggers and swords",
    "name": "Thief",
  },
}
`;
